File: rust/src/endpoints/freja.rs
use crate::client::{IdkollenError, PollOptions, WaitError};
use crate::models::{AgeVerificationRequest, AgeVerificationStatus};
use crate::models::{
    FrejaAuthRequest, FrejaBackchannelAuthRequest, FrejaBackchannelSignRequest, FrejaSignRequest,
    FrejaStatus,
};
use std::time::Instant;

#[cfg(feature = "async")]
use crate::client::IdkollenClient;

#[cfg(feature = "async")]
/// Freja eID operations.
pub struct FrejaEndpoint<'a>(pub(crate) &'a IdkollenClient);

#[cfg(feature = "async")]
impl FrejaEndpoint<'_> {
    /// Start a Freja eID authentication session.
    pub async fn auth(&self, req: FrejaAuthRequest) -> Result<FrejaStatus, IdkollenError> {
        self.0.post("/v3/freja/auth", &req).await
    }

    /// Start a Freja eID backchannel authentication session.
    pub async fn backchannel_auth(
        &self,
        req: FrejaBackchannelAuthRequest,
    ) -> Result<FrejaStatus, IdkollenError> {
        self.0.post("/v3/freja/backchannel/auth", &req).await
    }

    /// Start a Freja eID signing session.
    pub async fn sign(&self, req: FrejaSignRequest) -> Result<FrejaStatus, IdkollenError> {
        self.0.post("/v3/freja/sign", &req).await
    }

    /// Start a Freja eID backchannel signing session.
    pub async fn backchannel_sign(
        &self,
        req: FrejaBackchannelSignRequest,
    ) -> Result<FrejaStatus, IdkollenError> {
        self.0.post("/v3/freja/backchannel/sign", &req).await
    }

    /// Start a Freja eID age verification session.
    pub async fn age_verification(
        &self,
        req: AgeVerificationRequest,
    ) -> Result<AgeVerificationStatus, IdkollenError> {
        self.0.post("/v3/freja/age-verification", &req).await
    }

    /// Poll the current status of a Freja eID authentication session.
    pub async fn auth_status(&self, id: &str) -> Result<FrejaStatus, IdkollenError> {
        self.0.get(&format!("/v3/freja/auth/{}", id)).await
    }

    /// Poll the current status of a Freja eID signing session.
    pub async fn sign_status(&self, id: &str) -> Result<FrejaStatus, IdkollenError> {
        self.0.get(&format!("/v3/freja/sign/{}", id)).await
    }

    /// Poll the current status of a Freja eID age verification session.
    pub async fn age_verification_status(
        &self,
        id: &str,
    ) -> Result<AgeVerificationStatus, IdkollenError> {
        self.0
            .get(&format!("/v3/freja/age-verification/{}", id))
            .await
    }

    /// Cancel a Freja eID authentication session.
    pub async fn cancel_auth(&self, id: &str) -> Result<(), IdkollenError> {
        self.0.delete(&format!("/v3/freja/auth/{}", id)).await
    }

    /// Cancel a Freja eID signing session.
    pub async fn cancel_sign(&self, id: &str) -> Result<(), IdkollenError> {
        self.0.delete(&format!("/v3/freja/sign/{}", id)).await
    }

    /// Cancel a Freja eID age verification session.
    pub async fn cancel_age_verification(&self, id: &str) -> Result<(), IdkollenError> {
        self.0
            .delete(&format!("/v3/freja/age-verification/{}", id))
            .await
    }

    /// Poll until the authentication session reaches a terminal state or the timeout elapses.
    pub async fn wait_for_auth(
        &self,
        id: &str,
        opts: PollOptions,
    ) -> Result<FrejaStatus, WaitError> {
        let deadline = Instant::now() + opts.timeout;

        loop {
            let status = self.auth_status(id).await?;

            match status {
                FrejaStatus::Pending(_) => {
                    if Instant::now() >= deadline {
                        return Err(WaitError::Timeout);
                    }

                    tokio::time::sleep(opts.interval).await;
                },
                terminal => return Ok(terminal),
            }
        }
    }

    /// Poll until the signing session reaches a terminal state or the timeout elapses.
    pub async fn wait_for_sign(
        &self,
        id: &str,
        opts: PollOptions,
    ) -> Result<FrejaStatus, WaitError> {
        let deadline = Instant::now() + opts.timeout;

        loop {
            let status = self.sign_status(id).await?;

            match status {
                FrejaStatus::Pending(_) => {
                    if Instant::now() >= deadline {
                        return Err(WaitError::Timeout);
                    }

                    tokio::time::sleep(opts.interval).await;
                },
                terminal => return Ok(terminal),
            }
        }
    }

    /// Poll until the age verification session reaches a terminal state or the timeout elapses.
    pub async fn wait_for_age_verification(
        &self,
        id: &str,
        opts: PollOptions,
    ) -> Result<AgeVerificationStatus, WaitError> {
        let deadline = Instant::now() + opts.timeout;

        loop {
            let status = self.age_verification_status(id).await?;

            match status {
                AgeVerificationStatus::Pending(_) => {
                    if Instant::now() >= deadline {
                        return Err(WaitError::Timeout);
                    }

                    tokio::time::sleep(opts.interval).await;
                },
                terminal => return Ok(terminal),
            }
        }
    }
}

#[cfg(feature = "blocking")]
use crate::client::IdkollenBlockingClient;

#[cfg(feature = "blocking")]
/// Freja eID operations (blocking).
pub struct FrejaBlockingEndpoint<'a>(pub(crate) &'a IdkollenBlockingClient);

#[cfg(feature = "blocking")]
impl FrejaBlockingEndpoint<'_> {
    /// Start a Freja eID authentication session.
    pub fn auth(&self, req: FrejaAuthRequest) -> Result<FrejaStatus, IdkollenError> {
        self.0.post("/v3/freja/auth", &req)
    }

    /// Start a Freja eID backchannel authentication session.
    pub fn backchannel_auth(
        &self,
        req: FrejaBackchannelAuthRequest,
    ) -> Result<FrejaStatus, IdkollenError> {
        self.0.post("/v3/freja/backchannel/auth", &req)
    }

    /// Start a Freja eID signing session.
    pub fn sign(&self, req: FrejaSignRequest) -> Result<FrejaStatus, IdkollenError> {
        self.0.post("/v3/freja/sign", &req)
    }

    /// Start a Freja eID backchannel signing session.
    pub fn backchannel_sign(
        &self,
        req: FrejaBackchannelSignRequest,
    ) -> Result<FrejaStatus, IdkollenError> {
        self.0.post("/v3/freja/backchannel/sign", &req)
    }

    /// Start a Freja eID age verification session.
    pub fn age_verification(
        &self,
        req: AgeVerificationRequest,
    ) -> Result<AgeVerificationStatus, IdkollenError> {
        self.0.post("/v3/freja/age-verification", &req)
    }

    /// Poll the current status of a Freja eID authentication session.
    pub fn auth_status(&self, id: &str) -> Result<FrejaStatus, IdkollenError> {
        self.0.get(&format!("/v3/freja/auth/{}", id))
    }

    /// Poll the current status of a Freja eID signing session.
    pub fn sign_status(&self, id: &str) -> Result<FrejaStatus, IdkollenError> {
        self.0.get(&format!("/v3/freja/sign/{}", id))
    }

    /// Poll the current status of a Freja eID age verification session.
    pub fn age_verification_status(
        &self,
        id: &str,
    ) -> Result<AgeVerificationStatus, IdkollenError> {
        self.0.get(&format!("/v3/freja/age-verification/{}", id))
    }

    /// Cancel a Freja eID authentication session.
    pub fn cancel_auth(&self, id: &str) -> Result<(), IdkollenError> {
        self.0.delete(&format!("/v3/freja/auth/{}", id))
    }

    /// Cancel a Freja eID signing session.
    pub fn cancel_sign(&self, id: &str) -> Result<(), IdkollenError> {
        self.0.delete(&format!("/v3/freja/sign/{}", id))
    }

    /// Cancel a Freja eID age verification session.
    pub fn cancel_age_verification(&self, id: &str) -> Result<(), IdkollenError> {
        self.0.delete(&format!("/v3/freja/age-verification/{}", id))
    }

    /// Poll until the authentication session reaches a terminal state or the timeout elapses.
    pub fn wait_for_auth(&self, id: &str, opts: PollOptions) -> Result<FrejaStatus, WaitError> {
        let deadline = Instant::now() + opts.timeout;

        loop {
            let status = self.auth_status(id)?;

            match status {
                FrejaStatus::Pending(_) => {
                    if Instant::now() >= deadline {
                        return Err(WaitError::Timeout);
                    }

                    std::thread::sleep(opts.interval);
                },
                terminal => return Ok(terminal),
            }
        }
    }

    /// Poll until the signing session reaches a terminal state or the timeout elapses.
    pub fn wait_for_sign(&self, id: &str, opts: PollOptions) -> Result<FrejaStatus, WaitError> {
        let deadline = Instant::now() + opts.timeout;

        loop {
            let status = self.sign_status(id)?;

            match status {
                FrejaStatus::Pending(_) => {
                    if Instant::now() >= deadline {
                        return Err(WaitError::Timeout);
                    }

                    std::thread::sleep(opts.interval);
                },
                terminal => return Ok(terminal),
            }
        }
    }

    /// Poll until the age verification session reaches a terminal state or the timeout elapses.
    pub fn wait_for_age_verification(
        &self,
        id: &str,
        opts: PollOptions,
    ) -> Result<AgeVerificationStatus, WaitError> {
        let deadline = Instant::now() + opts.timeout;

        loop {
            let status = self.age_verification_status(id)?;

            match status {
                AgeVerificationStatus::Pending(_) => {
                    if Instant::now() >= deadline {
                        return Err(WaitError::Timeout);
                    }

                    std::thread::sleep(opts.interval);
                },
                terminal => return Ok(terminal),
            }
        }
    }
}

File: js/src/endpoints/freja.ts
import { IdkollenClient, IdkollenError, PollOptions } from "@/client";
import type {
  FrejaAuthRequest,
  FrejaBackchannelAuthRequest,
  FrejaBackchannelSignRequest,
  FrejaSignRequest,
  FrejaStatus,
} from "@/models/freja";
import type { AgeVerificationRequest, AgeVerificationStatus } from "@/models/age_verification";

const sleep = (ms: number): Promise<void> => new Promise((r) => setTimeout(r, ms));

/** Freja eID operations. */
export class FrejaEndpoint {
  constructor(private readonly client: IdkollenClient) {}

  /** Start a Freja eID authentication session. */
  public async auth(req: FrejaAuthRequest): Promise<FrejaStatus> {
    return this.client._post("/v3/freja/auth", req);
  }

  /** Start a Freja eID backchannel authentication session. */
  public async backchannelAuth(req: FrejaBackchannelAuthRequest): Promise<FrejaStatus> {
    return this.client._post("/v3/freja/backchannel/auth", req);
  }

  /** Start a Freja eID signing session. */
  public async sign(req: FrejaSignRequest): Promise<FrejaStatus> {
    return this.client._post("/v3/freja/sign", req);
  }

  /** Start a Freja eID backchannel signing session. */
  public async backchannelSign(req: FrejaBackchannelSignRequest): Promise<FrejaStatus> {
    return this.client._post("/v3/freja/backchannel/sign", req);
  }

  /** Poll the current status of a Freja eID authentication session. */
  public async authStatus(id: string): Promise<FrejaStatus> {
    return this.client._get(`/v3/freja/auth/${id}`);
  }

  /** Poll the current status of a Freja eID signing session. */
  public async signStatus(id: string): Promise<FrejaStatus> {
    return this.client._get(`/v3/freja/sign/${id}`);
  }

  /** Cancel a Freja eID authentication session. */
  public async cancelAuth(id: string): Promise<void> {
    return this.client._delete(`/v3/freja/auth/${id}`);
  }

  /** Cancel a Freja eID signing session. */
  public async cancelSign(id: string): Promise<void> {
    return this.client._delete(`/v3/freja/sign/${id}`);
  }

  /** Start a Freja eID age verification session. */
  public async ageVerification(req: AgeVerificationRequest): Promise<AgeVerificationStatus> {
    return this.client._post("/v3/freja/age-verification", req);
  }

  /** Poll the current status of a Freja eID age verification session. */
  public async ageVerificationStatus(id: string): Promise<AgeVerificationStatus> {
    return this.client._get(`/v3/freja/age-verification/${id}`);
  }

  /** Cancel a Freja eID age verification session. */
  public async cancelAgeVerification(id: string): Promise<void> {
    return this.client._delete(`/v3/freja/age-verification/${id}`);
  }

  /**
   * Poll until the authentication session reaches a terminal state or the timeout elapses.
   * @throws {IdkollenError} with code `"poll_timeout"` if the timeout is exceeded.
   */
  public async waitForAuth(
    id: string,
    opts: PollOptions = new PollOptions(),
  ): Promise<FrejaStatus> {
    const deadline = Date.now() + opts.timeoutMs;

    while (true) {
      const status = await this.authStatus(id);

      if (status.status !== "PENDING") {
        return status;
      }

      if (Date.now() >= deadline) {
        throw new IdkollenError("poll_timeout", 0, "Poll timed out");
      }

      await sleep(opts.intervalMs);
    }
  }

  /**
   * Poll until the signing session reaches a terminal state or the timeout elapses.
   * @throws {IdkollenError} with code `"poll_timeout"` if the timeout is exceeded.
   */
  public async waitForSign(
    id: string,
    opts: PollOptions = new PollOptions(),
  ): Promise<FrejaStatus> {
    const deadline = Date.now() + opts.timeoutMs;

    while (true) {
      const status = await this.signStatus(id);

      if (status.status !== "PENDING") {
        return status;
      }

      if (Date.now() >= deadline) {
        throw new IdkollenError("poll_timeout", 0, "Poll timed out");
      }

      await sleep(opts.intervalMs);
    }
  }

  /**
   * Poll until the age verification session reaches a terminal state or the timeout elapses.
   * @throws {IdkollenError} with code `"poll_timeout"` if the timeout is exceeded.
   */
  public async waitForAgeVerification(
    id: string,
    opts: PollOptions = new PollOptions(),
  ): Promise<AgeVerificationStatus> {
    const deadline = Date.now() + opts.timeoutMs;

    while (true) {
      const status = await this.ageVerificationStatus(id);

      if (status.status !== "PENDING") {
        return status;
      }

      if (Date.now() >= deadline) {
        throw new IdkollenError("poll_timeout", 0, "Poll timed out");
      }

      await sleep(opts.intervalMs);
    }
  }
}
